src/ZenohUTransport.cpp

```diff
@@ -10,3 +10,356 @@
+	}
+	std::string payload(sample.get_payload().as_string_view());
+	message.set_payload(payload);
+
+	return message;
+}
+
+ZenohUTransport::ZenohUTransport(const UUri& defaultUri,
+                                 const std::filesystem::path& configFile)
+    : UTransport(defaultUri),
+      session_(expect<Session>(open(
+          std::move(expect(config_from_file(configFile.string().c_str())))))) {}
+
+UStatus ZenohUTransport::registerRequestListener_(const std::string& zenoh_key,
+                                                  CallableConn listener) {
+	// NOTE: listener is captured by copy here so that it does not go out
+	// of scope when this function returns.
+	auto on_query = [this, listener](const Query& query) {
+		UAttributes attributes;
+		if (query.get_attachment().check()) {
+			attributes = attachmentToUAttributes(query.get_attachment());
+		}
+		auto id_str = serializer::uuid::AsString().serialize(attributes.id());
+		std::unique_lock<std::mutex> lock(query_map_mutex_);
+		query_map_.insert(std::make_pair(
+		    std::move(id_str), std::move(std::make_shared<OwnedQuery>(query))));
+	};
+
+	auto on_drop_queryable = []() {};
+
+	auto queryable = expect<Queryable>(
+	    session_.declare_queryable(zenoh_key, {on_query, on_drop_queryable}));
+
+	return UStatus();
+}
+
+UStatus ZenohUTransport::registerResponseListener_(const std::string& zenoh_key,
+                                                   CallableConn listener) {
+	std::unique_lock<std::mutex> lock(rpc_callback_map_mutex_);
+	rpc_callback_map_.insert(std::make_pair(zenoh_key, listener));
+
+	return UStatus();
+}
+
+UStatus ZenohUTransport::registerPublishNotificationListener_(
+    const std::string& zenoh_key, CallableConn listener) {
+	// NOTE: listener is captured by copy here so that it does not go out
+	// of scope when this function returns.
+	auto data_handler = [this, listener](const Sample& sample) mutable {
+		listener(sampleToUMessage(sample));
+		// invoke_nonblock_callback(&cb_sender, &listener_cloned, Ok(msg));
+	};
+
+	auto key = ListenerKey(listener, zenoh_key);
+	auto subscriber = expect<Subscriber>(
+	    session_.declare_subscriber(zenoh_key, data_handler));
+	{
+		std::unique_lock<std::mutex> lock(subscriber_map_mutex_);
+		subscriber_map_.insert(
+		    std::make_pair(std::move(key), std::move(subscriber)));
+	}
+	return UStatus();
+}
+
+UStatus ZenohUTransport::sendRequest_(const std::string& zenoh_key,
+                                      const std::string& payload,
+                                      const UAttributes& attributes) {
+	auto source_str =
+	    serializer::uri::AsString().serialize(attributes.source());
+	CallableConn resp_callback;
+	{
+		std::unique_lock<std::mutex> lock(rpc_callback_map_mutex_);
+
+		if (auto resp_callback_it = rpc_callback_map_.find(source_str);
+		    resp_callback_it == rpc_callback_map_.end()) {
+			return uError(UCode::UNAVAILABLE, "failed to find UUID");
+		} else {
```

Comment:
In relation to my prior comment about making a collection of thread safe data struct wrappers, this code shows the error object being created inside a locked region when it really shouldn't be. It seems better to me to have a thread-safe RpcCallbackMap class with an accessor like "[[nodiscard]] bool find(CallableConn& thing_to_capture)".